server/embed/serve.go

```diff
@@ -170,12 +175,22 @@ func (sctx *serveCtx) serve(
 		if tlsErr != nil {
 			return tlsErr
 		}
-		gs = v3rpc.Server(s, tlscfg, nil, gopts...)
+
+		gs := v3rpc.Server(s, tlscfg, nil, gopts...)
 		v3electionpb.RegisterElectionServer(gs, servElection)
 		v3lockpb.RegisterLockServer(gs, servLock)
 		if sctx.serviceRegister != nil {
 			sctx.serviceRegister(gs)
 		}
+
+		defer func(gs *grpc.Server) {
+			if err != nil {
+				sctx.lg.Warn("stopping secure grpc server due to error", zap.Error(err))
+				gs.Stop()
+				sctx.lg.Warn("stopped secure grpc server due to error", zap.Error(err))
+			}
+		}(gs)
```

Comment:
Minor comment: the code block is almost identical to the code block for insecure. I am thinking to define a separate function something like below, then call it in both secure and insecure code block. 

```
func closegRPCServer(gs *grpc.Server) {
  ...
}

```

Reply:
Since it needs `return error` to the log, I'm not sure how to handle this with your idea. In here, I follow the defer pattern~